File: OS LAB/Scan.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    vector<int> req = {98,183,37,122,14,124,65,67};
    int head = 53, disk_size = 200;
    string dir = "left";

    vector<int> left, right;
    int movement = 0;

    // split requests
    for (int r : req) {
        if (r < head) left.push_back(r);
        else if (r > head) right.push_back(r);
    }

    // sort
    sort(left.begin(), left.end(), greater<int>());
    sort(right.begin(), right.end());

    cout << "Order: ";

    if (dir == "left") {
        // move left
        for (int r : left) {
            cout << r << " ";
            movement += abs(head - r);
            head = r;
        }

        // go to 0
        movement += head;
        head = 0;

        // then right
        for (int r : right) {
            cout << r << " ";
            movement += abs(head - r);
            head = r;
        }
    } else {
        // move right
        for (int r : right) {
            cout << r << " ";
            movement += abs(head - r);
            head = r;
        }

        // go to end
        movement += (disk_size - 1 - head);
        head = disk_size - 1;

        // then left
        for (int r : left) {
            cout << r << " ";
            movement += abs(head - r);
            head = r;
        }
    }

    cout << "\nTotal movement: " << movement;
}

File: OS LAB/Disk Schedule Python/SCAN.py
def scan(requests, head, disk_size, direction):
    requests.sort()
    left = [r for r in requests if r < head]
    right = [r for r in requests if r >= head]

    path = []
    distance = 0

    if direction == "left":
        left.reverse()
        path.extend(left)
        path.append(0)
        path.extend(right)
    else:
        path.extend(right)
        path.append(disk_size - 1)
        left.reverse()
        path.extend(left)

    current = head
    for r in path:
        distance += abs(current - r)
        current = r

    print("SCAN Path:", [head] + path)
    print("Total Distance:", distance)


# Example
scan([98,183,37,122,14,124,65,67], 53, 200, "right")